LogMonitor/src/LogMonitor/FileMonitor/Utilities.cpp

```diff
@@ -81,82 +82,131 @@ HANDLE GetLogDirHandle(std::wstring logDirectory, HANDLE stopEvent) {
-                                          OPEN_EXISTING,
-                                          FILE_FLAG_BACKUP_SEMANTICS | FILE_FLAG_OVERLAPPED,
-                                          nullptr);
-
-            if (logDirHandle == INVALID_HANDLE_VALUE)
-            {
-                status = GetLastError();
-                ++retryCount;
-            }
-            else
-            {
-                status = ERROR_SUCCESS;
-                break;
-            }
+            status = GetLastError();
+            logWriter.TraceError(
+                Utility::FormatString(
+                    L"Failed to set timer object to monitor log file changes in directory %s. Error: %lu",
+                    logDirectory.c_str(),
+                    status)
+                    .c_str());
+            break;
         }
 
-        CancelWaitableTimer(timerEvent);
-        CloseHandle(timerEvent);
+        DWORD wait = WaitForMultipleObjects(eventsCount, dirOpenEvents, FALSE, INFINITE);
+        switch (wait)
+        {
+        case WAIT_OBJECT_0:
+        {
+            //
+            // The process is exiting. Stop the timer and return.
+            //
+            CancelWaitableTimer(timerEvent);
+            CloseHandle(timerEvent);
+            return INVALID_HANDLE_VALUE;
+        }
+
+        case WAIT_OBJECT_0 + 1:
+        {
+            //
+            // Timer event. Retry opening directory handle.
+            //
+            break;
+        }
+
+        default:
+        {
+            //
+            // Wait failed, return the failure.
+            //
+            status = GetLastError();
+
+            CancelWaitableTimer(timerEvent);
+            CloseHandle(timerEvent);
+
+            return INVALID_HANDLE_VALUE;
+        }
+        }
+
+        logDirHandle = CreateFileW(logDirectory.c_str(),
+                                   FILE_LIST_DIRECTORY,
+                                   FILE_SHARE_READ | FILE_SHARE_WRITE,
+                                   nullptr,
+                                   OPEN_EXISTING,
+                                   FILE_FLAG_BACKUP_SEMANTICS | FILE_FLAG_OVERLAPPED,
+                                   nullptr);
+
+        if (logDirHandle == INVALID_HANDLE_VALUE)
+        {
+            status = GetLastError();
+            elapsedTime += WAIT_INTERVAL;
+        }
+        else
+        {
+            status = ERROR_SUCCESS;
+            break;
+        }
     }
 
     return logDirHandle;
 }
+
+// Converts the time to wait to a large integer
+LARGE_INTEGER _ConvertWaitIntervalToLargeInt(int timeInterval)
+{
+    LARGE_INTEGER liDueTime{};
+
+    int millisecondsToWait = timeInterval * 1000;
+    liDueTime.QuadPart = -millisecondsToWait * 10000LL; // wait time in 100 nanoseconds
+    return liDueTime;
+}
+
+// Returns the time (in seconds) to wait based on the specified waitInSeconds
+int _GetWaitInterval(std::double_t waitInSeconds, int elapsedTime)
```

Comment:
Why is wait in seconds a double? Does it support fraction of a second waits?

Reply:
![image](https://github.com/microsoft/windows-container-tools/assets/20437227/2e4d3ef3-9287-402b-a904-8eeabf3be768)

`int` type has a finite range of values that it can represent. We therefore have `waitInSeconds` as a double to avoid in `INFINITY` being converted to an integer.